src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -940,26 +970,74 @@ where
                     let batch = batch
                         .borrow_mut()
                         .take()
                         .expect("batch already taken, but we should be the only consumer");
                     for (pid, offset) in batch.source_upper.iter() {
-                        let previous_offset = global_source_upper.insert(pid.clone(), *offset);
-                        if let Some(previous_offset) = previous_offset {
-                            assert!(previous_offset <= *offset);
+                        // If we have a time for the batch in `batch_times` already, we go onto
+                        // checking the frontier. Otherwise, place the time and offset in the map
+                        // and wait for the next empty batch to push us along.
+                        if let Some((batch_ts, advance_to)) = batch_times.get(pid) {
+                            trace!(
+                                "reclock({id}) {worker_id}/{worker_count}: batch frontier: {:?}, \
+                                processing batch at ts: {}, \
+                                wanting to advance to: {:?}",
+                                frontiers[0].frontier(),
+                                batch_ts,
+                                advance_to
+                            );
+                            // If the batch_input frontier has advanced past this batch time, then
+                            // we can advance the `global_source_upper`, which will later downgrade
+                            // `cap_set`, as we are sure we have seen all messages for this
+                            // `source_upper`.
+                            let should_advance =
+                                if let Some(frontier_ts) = frontiers[0].frontier().as_option() {
+                                    frontier_ts > batch_ts
+                                } else {
+                                    // The empty frontier is trivially past all times
+                                    true
+                                };
+
+                            if should_advance {
+                                global_source_upper.insert(pid.clone(), *advance_to);
+                                // We also replace the value in `batch_times` with the new
+                                // offset, associated with the frontier, to move it along
+                                batch_times.insert(pid.clone(), (cap.time().clone(), *offset));
+                                continue;
+                            } else if offset > advance_to {
+                                // Otherwise we advance to the higher offset, assuming
+                                // that eventually the batch frontier will eventually
+                                // go past this time.
+                                batch_times.insert(pid.clone(), (cap.time().clone(), *offset));
+                            }
```

Comment:
subtley question 4: will this always make progress? or do I need to make the `HashMap` hold a list of `(Timestamp, Offset)` pairs? I think their MAY be a pathological operator scheduling that could get this stuck forever, but I could be misinterpretting? The idea would be that the frontier and new batches just past that frontier are always coming in at the same time?